testutil/filefixtures/types.go

```diff
@@ -0,0 +1,184 @@
+	}
+	if f.Metadata.Gid == 0 {
+		f.Metadata.Gid = 10000
+	}
+	if f.Metadata.ModTime.IsZero() {
+		f.Metadata.ModTime = time.Unix(0, 0).UTC()
+	}
+	if f.Metadata.AccessTime.IsZero() {
+		f.Metadata.AccessTime = time.Unix(0, 0).UTC()
+	}
+	return f
+}
+
+/*
+	Create files described by the fixtures on the real filesystem path given.
+*/
+func (ffs Fixture) Create(basePath string) {
+	for _, f := range ffs.Files {
+		fs.PlaceFile(basePath, f.Metadata, bytes.NewBuffer(f.Body))
+	}
+	// re-do time enforcement... in reverse order, so we cover our own tracks
+	for i := len(ffs.Files) - 1; i >= 0; i-- {
+		f := ffs.Files[i]
+		if f.Metadata.Typeflag == tar.TypeDir {
+			fs.PlaceDirTime(basePath, f.Metadata)
+		}
+	}
+}
+
+/*
+	Scan a real filesystem and see it as fixture file descriptions.
+	Usually used as a prelude to a pair of `Describe` calls followed by
+	an equality assertion.
+
+	Note that this loads all file bodies into memory at once, so it
+	is not wise to use on large filesystems.
+
+	Result will be sorted by filename, as per usual.
+*/
+func Scan(basePath string) Fixture {
+	ffs := Fixture{fmt.Sprintf("Scan of %q", basePath), nil}
+	preVisit := func(filenode *fs.FilewalkNode) error {
+		if filenode.Err != nil {
+			return filenode.Err
+		}
+		hdr, file := fs.ScanFile(basePath, filenode.Path)
+		var body []byte
+		if file != nil {
+			defer file.Close()
+			var err error
+			body, err = ioutil.ReadAll(file)
+			if err != nil {
+				return err
+			}
+		}
+		ffs.Files = append(ffs.Files, FixtureFile{hdr, body})
+		return nil
+	}
+	if err := fs.Walk(basePath, preVisit, nil); err != nil {
+		panic(err)
+	}
+	sort.Sort(filesByPath(ffs.Files))
+	return ffs
+}
+
+/*
+	Produces a string where every line describes a one entry in the
+	set of file descriptions.  This is useful for handing into "ShouldResemble"
+	assertions, which will not only pass/fail but do character diffs which
+	effectively cover the whole structure in one shot.
+
+
+*/
+func (ffs Fixture) Describe(opts ComparisonOptions) string {
+	lines := make([]string, len(ffs.Files))
+	for i, f := range ffs.Files {
+		lines[i] = f.Describe(opts)
+	}
+	return strings.Join(lines, "\n")
+}
+
+/*
+	As per `FixtureFiles.Describe`, but this is for a single entry.
+*/
+func (ff FixtureFile) Describe(opts ComparisonOptions) string {
+	parts := []struct {
+		Key   string
+		Value interface{}
+	}{
+		{"Name:%q", ff.Metadata.Name},
+		{"Type:%q", ff.Metadata.Typeflag},
+		{"Perms:%q", (map[bool]interface{}{true: "-", false: ff.Metadata.FileMode()})[opts&ComparePerms == 0]},
+		{"Mtime:%q", (map[bool]interface{}{true: "-", false: ff.Metadata.ModTime.UTC()})[opts&CompareMtime == 0]},
+		{"Atime:%q", (map[bool]interface{}{true: "-", false: ff.Metadata.AccessTime.UTC()})[opts&CompareAtime == 0]},
+		{"Uid:%d", (map[bool]interface{}{true: "-", false: ff.Metadata.Uid})[opts&CompareUid == 0]},
+		{"Gid:%d", (map[bool]interface{}{true: "-", false: ff.Metadata.Gid})[opts&CompareGid == 0]},
```

Comment:
TERRRRRRRRRNARRRYS.

It was way worse before: https://github.com/polydawn/repeatr/commit/3897dce9ea2fb5a3e09f7daaf8e5ef918c1ca427